### PR TITLE
[Feature](multi-catalog) Add memory tracker for orc reader/writer and arrow parquet writer。

### PR DESCRIPTION
**Pull request by kaka11chen**
## Proposed changes

backport #37234

### CODE REVIEW

**Comment by kaka11chen** [comment]
run buildall

**Comment by zgxme** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

File: be/src/vec/common/allocator.h [hunk 1]
```diff
@@ -112,47 +251,66 @@ class Allocator {
...
         return buf;
     }
 
     /// Free memory range.
     void free(void* buf, size_t size) {
+        /*size_t record_size = 0;
+        if (size) {
+            record_size = size;
+        } else {
+            record_size = MemoryAllocator::allocated_size(buf);
```
Comment (xinyiZzz) [inline be/src/vec/common/allocator.h:299]: remove this commented codes

File: be/src/vec/common/allocator.h [hunk 2]
```diff
@@ -68,6 +69,143 @@ static constexpr size_t MALLOC_MIN_ALIGNMENT = 8;
...
+private:
+    static std::unordered_map<void*, size_t> _allocated_sizes;
+    static std::mutex _mutex;
+};
+
+#if defined(USE_JEMALLOC)
+#include <jemalloc/jemalloc.h>
+class ArrowJemallocMemoryAllocator {
+public:
+    static void* malloc(size_t size) __THROW { return je_malloc(size); }
```
Comment (xinyiZzz) [inline be/src/vec/common/allocator.h:189]: `je_xxx`, which Jemalloc is actually used?
Doris Jemalloc has the prefix `je`, without an underscore

File: be/src/vec/common/allocator.h [hunk 3]
```diff
@@ -68,6 +69,143 @@ static constexpr size_t MALLOC_MIN_ALIGNMENT = 8;
...
+
+    static int posix_memalign(void** ptr, size_t alignment, size_t size) __THROW {
+        return je_posix_memalign(ptr, alignment, size);
+    }
+
+    static void* realloc(void* ptr, size_t size) __THROW { return je_realloc(ptr, size); }
+
+    static void free(void* p) __THROW { je_free(p); }
+
+    static void release_unused() {
```
Comment (xinyiZzz) [inline be/src/vec/common/allocator.h:203]: seems that `ArrowJemallocMemoryAllocator` only adds `release_unused` compared to `DefaultMemoryAllocator`

File: be/src/vec/common/allocator.h (hunk 2 above)
Comment (xinyiZzz) [reply be/src/vec/common/allocator.h:189]: img_v3_02d2_b63c11c9-8a19-429e-a981-f1dffe1c2e6g
`je_malloc` is `jemalloc`

File: be/src/vec/exec/format/parquet/arrow_memory_pool.h [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            buffer[0] = kDeallocPoison;
+            buffer[size - 1] = kDeallocPoison;
+        }
+#endif
+        _allocator.deallocate_aligned(buffer, size, alignment);
+
+        _stats.did_free_bytes(size);
+    }
+
+    void ReleaseUnused() override { _allocator.release_unused(); }
```
Comment (xinyiZzz) [inline be/src/vec/exec/format/parquet/arrow_memory_pool.h:187]: `je_mallctl(fmt::format("arena.{}.purge", MALLCTL_ARENAS_ALL)` is a very heavy and time-consuming operation.

if jemalloc dirty pages exceeding GB, this method may take more than 1 second to clear.
so be careful not to block the program.

Comment (kaka11chen) [reply be/src/vec/exec/format/parquet/arrow_memory_pool.h:182]: > `je_mallctl(fmt::format("arena.{}.purge", MALLCTL_ARENAS_ALL)` is a very heavy and time-consuming operation.
> 
> if jemalloc dirty pages exceeding GB, this method may take more than 1 second to clear. so be careful not to block the program.

Default arrow memory pool also defined this function.

File: be/src/vec/common/allocator.h (hunk 3 above)
Comment (kaka11chen) [reply be/src/vec/common/allocator.h]: Will change to `DefaultMemoryAllocator`.

File: be/src/vec/common/allocator.h (hunk 2 above)
Comment (kaka11chen) [reply be/src/vec/common/allocator.h]: Will Change to `DefaultMemoryAllocator`.

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Review by xinyiZzz — APPROVED** [review]
LGTM

File: be/src/vec/common/allocator.h [hunk 5]
```diff
@@ -112,32 +232,42 @@ class Allocator {
...
-                int res = posix_memalign(&buf, alignment, size);
+                int res = MemoryAllocator::posix_memalign(&buf, alignment, size);
 
                 if (0 != res) {
                     release_memory(size);
                     throw_bad_alloc(
                             fmt::format("Cannot allocate memory (posix_memalign) {}.", size));
                 }
-
-                if constexpr (clear_memory) memset(buf, 0, size);
```
Comment (yiguolei) [inline be/src/vec/common/allocator.h:138]: why delete clear memory？？

**Comment by kaka11chen** [comment]
run buildall

**Comment by kaka11chen** [comment]
run buildall

**Pull request merged**